### PR TITLE
Deprecate DTDConnectionInfo and replace with a class that can handle separate private/exposed URIs

### PR DESCRIPTION
**Pull request by DanTup**
This replaces `DTDConnectionInfo` with `DTDInfo` (unsure if a better name?) and marks the former as deprecated.

The new class has two URIs:

- `localUri` which is the URI used for backend services on the same machine as the DTD process (this is usually a `localhost` URI)
- `exposedUri` which is the URI that can be used from the frontend where the user is (this may or may not be the same as localUri depending on whether they're in a remote/web IDE session)

The DevTools web API returns the exposed URI (since it will be used by frontend tools) and other places that used the URI for connecting directly use the local one.

On its own, this change doesn't do anything, the intention is to roll this into the SDK and update the DevTools server code in dartdev/DDS to accept an additional optional URI which will then provide these back to the server code here (for the web API handler and extension search code).

This is work towards fixing some of the issues in https://github.com/Dart-Code/Dart-Code/issues/5158.

### CODE REVIEW

**Comment by DanTup** [comment]
@kenzieschmoll is this something that would be included in the release notes? (it doesn't really affect the DevTools app - at least not without other fixes - only devtools_shared users).

While I left the old typedef and made it deprecated, these changes are still breaking for any code that calls any of these methods (such as the handlers) that have had their signatures changed (this means that rolling the new devtools_shared into the SDK will also require some code changes, which I will start preparing a CL for).

**Comment by DanTup** [comment]
Obviously it can't be landed yet, but I prepared the change for the SDK that will be needed to roll this change into the SDK:

https://dart-review.googlesource.com/c/sdk/+/383400

It does not add support for passing a second URI yet, it's just the minimum changes to the APIs that would keep things working as-is with this change (I don't want to complicate rolling a new devtools_shared into the SDK further). I'll add the additional flag for passing another URI in another change after this is done.

**Review by kenzieschmoll — APPROVED** [review]

File: packages/devtools_shared/lib/src/common.dart [hunk 1]
```diff
@@ -3,4 +3,37 @@
 // found in the LICENSE file.
 
 /// Describes an instance of the Dart Tooling Daemon.
+@Deprecated('Use DTDInfo instead')
 typedef DTDConnectionInfo = ({String? uri, String? secret});
+
+/// Information about a Dart Tooling Daemon instance.
+class DTDInfo {
```
[the remark this answers is not in the document]
Comment (DanTup) [reply packages/devtools_shared/lib/src/common.dart]: Done (40d908a4f) - I never know what the conventions are for acronyms like this so just copied the previous typedef. I wonder if there's a lint for strings of uppercase chars if we shouldn't ever do that 😄

File: packages/devtools_shared/CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,7 @@
 * Deprecate `surveyActionTakenPropertyName`.
 * Deprecate `apiGetSurveyShownCount` in favor of `SurveyApi.getSurveyShownCount`.
 * Deprecate `apiIncrementSurveyShownCount` in favor of `SurveyApi.incrementSurveyShownCount`.
+* Deprecate `DTDConnectionInfo` in favor of `DTDInfo` which supports tracking to URIs for DTD to better support web/remote environments.
```
[the remark this answers is not in the document]
Comment (DanTup) [reply packages/devtools_shared/CHANGELOG.md]: Thanks! Fixed in b04192741

File: packages/devtools_shared/lib/src/common.dart [hunk 3]
```diff
@@ -3,4 +3,37 @@
...
+    Uri? exposedUri,
+    this.secret,
+  }) : exposedUri = exposedUri ?? localUri;
+
+  /// The URI for connecting to DTD from the backend.
+  ///
+  /// This is usually a `http://localhost/` address that is accessible to tools
+  /// running in the same location as the DTD process. It may NOT be accessible
+  /// to frontends that may run in another location - for example the DevTools
+  /// frontend running in a browser (or VS Code UI) in a remote/web IDE session.
```
[the remark this answers is not in the document]
Comment (DanTup) [reply packages/devtools_shared/lib/src/common.dart]: Also in b04192741

File: packages/devtools_shared/lib/src/server/handlers/_dtd.dart [hunk 4]
```diff
@@ -9,10 +9,15 @@ part of '../server_api.dart';
...
-    DTDConnectionInfo? dtd,
+    DTDInfo? dtd,
   ) {
     return ServerApi._encodeResponse(
-      {DtdApi.uriPropertyName: dtd?.uri},
+      {
+        // Always provide the exposed URI to callers of the web API.
+        // TODO(dantup): Should we add properties for both URIs and deprecate
+        //  "uri"? Would anyone on the backend ever call this API?
+        DtdApi.uriPropertyName: dtd?.exposedUri,
```
[the remark this answers is not in the document]
Comment (DanTup) [reply packages/devtools_shared/lib/src/server/handlers/_dtd.dart]: sgtm, I've removed the TODO for now then (also b04192741)

File: packages/devtools_shared/CHANGELOG.md [hunk 5]
```diff
@@ -13,7 +13,7 @@
 * Deprecate `surveyActionTakenPropertyName`.
 * Deprecate `apiGetSurveyShownCount` in favor of `SurveyApi.getSurveyShownCount`.
 * Deprecate `apiIncrementSurveyShownCount` in favor of `SurveyApi.incrementSurveyShownCount`.
-* Deprecate `DTDConnectionInfo` in favor of `DTDInfo` which supports tracking to URIs for DTD to better support web/remote environments.
+* Deprecate `DTDConnectionInfo` in favor of `DTDInfo` which supports tracking two URIs for DTD to better support web/remote environments.
```
Comment (kenzieschmoll) [inline packages/devtools_shared/CHANGELOG.md:16]: r/DTDInfo/DtdInfo

**Review by kenzieschmoll — APPROVED** [review]

**Comment by kenzieschmoll** [comment]
Please merge with master to resolve the CI failures.

**Comment by DanTup** [comment]
Great, thanks!

**Comment by DanTup** [comment]
These failures look legit - we switched to using `Uri` instead of string, and some of the tests use invalid URIs that won't parse as URIs. I'll fix these up next week.

**Comment by DanTup** [comment]
When a new devtools_shared is released, we'll need to apply the changes in https://dart-review.googlesource.com/c/sdk/+/383400 while rolling it in to the SDK.

**Comment by kenzieschmoll** [comment]
devtools_shared 10.1.0 has been published, and you can use the `fbc2f1cda3809462811ce503197b8155df77dda6` commit hash to update the devtools rev in the DEPS file.

**Comment by DanTup** [comment]
Great, thanks!

**Comment by kenzieschmoll** [comment]
10.1.0 was retracted. Republishing as 11.0.0: https://github.com/flutter/devtools/pull/8313

**Comment by jakemac53** [comment]
> 10.1.0 was retracted. Republishing as 11.0.0: #8313

Awesome thanks!

**Pull request merged**